language/ast.py

```diff
@@ -0,0 +1,779 @@
+    """
+
+    element_type = Field(type=ID)
+
+
+@abstract
+class EnumerationDef(TypeDef):
+    """
+    Base class for enumeration
+    """
+
+
+class PositionalEnumerationDef(EnumerationDef):
+    """
+    Positional enumeration
+    """
+
+    elements = Field(type=UnqualifiedID.list)
+
+
+class ElementValueAssoc(RFLXNode):
+    """
+    Element/value association
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    literal = Field(type=NumericLiteral)
+
+
+class NamedEnumerationDef(EnumerationDef):
+    """
+    Named enumeration
+    """
+
+    elements = Field(type=ElementValueAssoc.list)
+
+
+class EnumerationTypeDef(TypeDef):
+    """
+    Enumeration type definition
+    """
+
+    elements = Field(type=EnumerationDef)
+    aspects = Field(type=Aspect.list)
+
+
+class Then(RFLXNode):
+    """
+    Link to field
+    """
+
+    target = Field(type=NullID)
+    aspects = Field(type=Aspect.list)
+    condition = Field(type=Expr)
+
+
+class NullComponent(RFLXNode):
+    """
+    Null message component
+    """
+
+    then = Field(type=Then)
+
+
+class Component(RFLXNode):
+    """
+    Message component
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    type_identifier = Field(type=ID)
+    aspects = Field(type=Aspect.list)
+    condition = Field(type=Expr)
+    thens = Field(type=Then.list)
+
+
+class Components(RFLXNode):
+    """
+    Message components
+    """
+
+    initial_component = Field(type=NullComponent)
+    components = Field(type=Component.list)
+
+
+@abstract
+class BaseChecksumVal(RFLXNode):
+    """
+    Base class for checksum values
+    """
+
+
+class ChecksumVal(BaseChecksumVal):
+    """
+    Single checksum value
+    """
+
+    data = Field(type=Expr)
+
+
+class ChecksumValueRange(BaseChecksumVal):
+    """
+    Checksum value range
+    """
+
+    lower = Field(type=Expr)
+    upper = Field(type=Expr)
+
+
+class ChecksumAssoc(RFLXNode):
+    """
+    Association between checksum field and list of covered fields
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    covered_fields = Field(type=BaseChecksumVal.list)
+
+
+class ChecksumAspect(RFLXNode):
+    """
+    Checksum aspect
+    """
+
+    associations = Field(type=ChecksumAssoc.list)
+
+
+class MessageTypeDef(AbstractMessageTypeDef):
+    """
+    Message type definition
+    """
+
+    components = Field(type=Components)
+    checksums = Field(type=ChecksumAspect)
+
+
+class UnqualifiedVariable(Expr):
+    """
+    Unqualified variable
+    """
+
+    identifier = Field(type=UnqualifiedID)
+
+
+class Variable(Expr):
+    """
+    Variable
+    """
+
+    identifier = Field(type=ID)
+
+
+@abstract
+class AttrBase(RFLXNode):
+    """
+    Base class for attributes
+    """
+
+
+class Attr(AttrBase):
+    """
+    Attribute
+    """
+
+    enum_node = True
+    alternatives = [
+        "First",
+        "Size",
+        "Last",
+        "Valid_Checksum",
+        "Head",
+        "Opaque",
+        "Present",
+        "Valid",
+    ]
```

Comment:
Is there any reason to define an abstract base class? From what I can see `Attr` would also include all attributes as there is no other child of `AttrBase`.

Reply:
Removed in 53ef203.